### ISSUE TITLE
Group Labels API

### ISSUE THREAD

**Issue by davidobber** [gitlab4j/gitlab4j-api#477]
Access to Group Labels API is missing, the API are almost identical to the Project Labels API . I've attached two possible implementations based on LabelsApi source.

[group-labels.zip](https://github.com/gitlab4j/gitlab4j-api/files/3857486/group-labels.zip)

**Comment by gmessner** [comment]
@davidobber 
Thanks for providing some alternative ways to do this.  We prefer to combine multiple APIs into a single class, and this one makes sense to put both the project and group labels API into ```LabelsApi```.  There has also been some significant updates to the actual APIs on GitLab, so I've created a branch that combines the two and made the necessary updates for the GitLab API changes.

The other thing we are now doing is not having multiple create and update methods, instead prefer to pass an object in which holds the properties, that class provides ```withXXXX()``` methods to set the properties.  Here is an example:
```
Label labelProperties = new Label()
            .withName("green-group-label")
            .withColor("#00FF00")
gitLabApi.getLabelsApi().createGroupLabel(projectId, labelProperties);
```

I have not merged this yet as there are no tests for it, I will be adding those tomorrow and then will do a release.

Please take a look at the branch and provide any feedback you may have.

**Comment by davidobber** [comment]
Ok, the LabelApi exposes all the methods related to labels, project & group
labels, as the class name suggests. The "with" approach it's a "builder
like" approach that works directly to the object and it works well as long
as it's a pojo, so the order of the "withXXX" methods used does not involve
validation rules. I imagine that the GitLabApiForm represents the json
body.
I tried to find a way to access to the Admin Labels, but at the moment I
didn't find the API to do that.





Il giorno mar 19 nov 2019 alle ore 20:41 Greg Messner <
notifications@github.com> ha scritto:

**Comment by gmessner** [comment]
@davidobber 
The ```withXXX()``` builder pattern is used extensively in this library and it does not effect validation of the model classes.  These are not necessarily POJOs  as some of the have some specific annotations related to Jackson JSON.

The API calls that do POST and PUT HTTP methods, the body is not JSON, it is application/x-www-form-urlencoded  (form data).

Concerning the admin labels, looks like they are only available in the UI, you might want to put a feature request in with GitLab for this to be added to the GitLab API, once added to the API we can add it to this library.

**Comment by gmessner** [comment]
@davidobber 
I've added integration tests and have released version 4.14.1.  Please verify and close this issue when you have time.

**Comment by davidobber** [comment]
What I meant is that it's not exactly the builder pattern, because you're
working directly on object to set it up, and it's perfectly fine. Using the
builder pattern as a builder means that a build() method at the end builds
the object, but before creating it, the method can check consistency
between the builder state params. It's not necessary as long as it's a
passthrough builder (in this case is just a waist of time). But there are
some cases where a setter can cause side effects or where the builder needs
to know everything to choose which object to create or which constructor to
invoke.

I will open a feature request for admin labels.







Il giorno mer 20 nov 2019 alle ore 04:26 Greg Messner <
notifications@github.com> ha scritto:

**Comment by gmessner** [comment]
@davidobber 
As you state it is not the builder pattern but it does check state of the POJO when getForm() is called, making it similar to the builder pattern.

**Comment by gmessner** [comment]
@davidobber Since the functionality was released in 4.14.1 I'm going to close this issue, If you find any problems with the LabelsApi please feel free to re-open this issue.
